### ISSUE TITLE
Typings for attach fails if specify effect argument type

### ISSUE THREAD

**Issue by PFight** [effector/effector#439]
**What is the current behavior:**

There exact sample from https://effector.dev/docs/api/effector/attach, but with one change - when calling createEffect I added type for effect argument:
```
         const original = createEffect<string, void>(params => {
            console.log('Original effect called with', params);
        });
        
        const data =createStore(8900);
        
        const created = attach({
            effect: original,
            source: data,
            mapParams: (params, data) => {
                console.log('Created effect called with', params, 'and data', data);
                return {wrapped: params, data};
            },
        });
```
TypeScript 3.8.0 compailns with error:

```
No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ effect: Effect<string, void, Error>; source: Store<number>; mapParams: (params: unknown, data: number) => { wrapped: unknown; data: number; }; }' is not assignable to parameter of type '{ effect: Effect<string, void, Error>; }'.
      Object literal may only specify known properties, and 'source' does not exist in type '{ effect: Effect<string, void, Error>; }'.
```

**Which versions of effector packages, and which browser and OS are affected by this issue? Did this work in previous versions of effector?:**

"effector": "^21.7.5",
TypeScript 3.8.0 

Maybe the last TypeScript will be ok, can't check right now.

**Comment by zerobias** [comment]
Looks like you have type error: mapParams returns `{wrapped: any; data: number}` but original effect accepts only string. Typescript prefer to look at mapParams at first, so its message rather out of place

However, explicit generics usually reveal actual messages:

```ts
const original = createEffect((params: string) => {
    console.log('Original effect called with', params)
  })

const data = createStore(8900)

const created = attach<number, typeof data, typeof original>({
  effect: original,
  source: data,
  mapParams: (params, data) => {
    console.log('Created effect called with', params, 'and data', data)
    return {wrapped: params, data}
  },
})
```

**Comment by PFight** [comment]
Hm, really it was error in typings. Quite dificult to determine error source until write explicit typings... Strange typescript behaviour...

Thanks.

**Comment by PFight** [comment]
I found, that problem with typing solves if rename overloads of attach with `{ source, effect, mapParams }` arguments to another function name. For example, call it attachToSource. Then typescript error messages become clear and usefull.

Ho about to extract such attachToSource function?

**Comment by zerobias** [comment]
Separate methods overwhelm users, there is a direct correlation between amount of methods in api and „this is a large api“ feeling, so I'd rather [try to improve](https://github.com/effector/effector/commit/2a17973e2949a86575b55ed6ea5bbd1b8a9eca85) inference for attach itself

**Comment by zerobias** [comment]
Good news! I found a way to show [useful error message](https://github.com/effector/effector/commit/008931099d0bfde758a8776047207e15daff1d8a#diff-09a8b3795eede343a226de3a8e5de2b5db257c69036aa66eea64bd2c62179a4eR198) in your case

Will be released in `effector 21.8.0`

**Comment by PFight** [comment]
Interesting solution, saved for myself. Thanks!
